File: RacePerformanceTest.java
import zio.*;
import cats.effect.IO;
import cats.effect.unsafe.implicits.global;

public class RacePerformanceTest {
    private static final int ITERATIONS = 10000;

    public static void main(String[] args) {
        System.out.println("Starting performance test...");
        System.out.println("Running " + ITERATIONS + " iterations for each implementation");

        // Warmup
        long catsWarmupTime = runCatsRace();
        System.out.println("Cats-Effect warmup time (ns): " + catsWarmupTime);

        long zioWarmupTime = runZioRace();
        System.out.println("ZIO warmup time (ns): " + zioWarmupTime);

        // Actual test
        long catsTime = runCatsRace();
        System.out.println("Cats-Effect time (ns): " + catsTime);

        long zioTime = runZioRace();
        System.out.println("ZIO time (ns): " + zioTime);

        double ratio = (double) catsTime / zioTime;
        System.out.println("Performance ratio: Cats-Effect / ZIO = " + ratio);
        System.out.println("ZIO is " + ratio + "x faster than Cats-Effect");
    }

    private static long runCatsRace() {
        long startTime = System.nanoTime();

        int result = loop(0);
        if (result != ITERATIONS) {
            throw new RuntimeException("Unexpected result: " + result);
        }

        long endTime = System.nanoTime();
        return endTime - startTime;
    }

    private static int loop(int i) {
        if (i < ITERATIONS) {
            return IO.race(IO.never(), IO.delay(() -> i + 1))
                    .flatMap(either -> either.fold(
                            left -> IO.pure(left), // This should never happen
                            right -> loop(right)
                    ))
                    .unsafeRunSync();
        } else {
            return i;
        }
    }

    private static long runZioRace() {
        long startTime = System.nanoTime();

        int result = Unsafe.unsafe(unsafe -> {
            return Runtime.default_().unsafe.run(
                    zioLoop(0)
            ).getOrThrowFiberFailure();
        });

        if (result != ITERATIONS) {
            throw new RuntimeException("Unexpected result: " + result);
        }

        long endTime = System.nanoTime();
        return endTime - startTime;
    }

    private static ZIO<Object, Object, Integer> zioLoop(int i) {
        if (i < ITERATIONS) {
            return ZIO.never().race(ZIO.succeed(i + 1))
                    .flatMap(either -> either.fold(
                            left -> ZIO.succeed(left), // This should never happen
                            right -> zioLoop(right)
                    ));
        } else {
            return ZIO.succeed(i);
        }
    }
}

File: RacePerformanceTest.scala
import zio._
import cats.effect.unsafe.implicits.global
import scala.concurrent.duration._
// Import the SimplifiedOptimizedRace implementation
import zio.SimplifiedOptimizedRace

object RacePerformanceTest extends ZIOAppDefault {
  val iterations = 10000

  def runCatsRace(): Long = {
    import cats.effect.IO
    
    val startTime = System.nanoTime()
    
    def loop(i: Int): IO[Int] =
      if (i < iterations) IO.race(IO.never, IO.delay(i + 1)).flatMap(_ => loop(i + 1))
      else IO.pure(i)
    
    loop(0).unsafeRunSync()
    
    val endTime = System.nanoTime()
    endTime - startTime
  }

  def runZioRace(): Long = {
    val startTime = System.nanoTime()
    
    def loop(i: Int): UIO[Int] =
      if (i < iterations) ZIO.never.race(ZIO.succeed(i + 1)).flatMap(_ => loop(i + 1))
      else ZIO.succeed(i)
    
    Unsafe.unsafe { implicit unsafe =>
      Runtime.default.unsafe.run(loop(0)).getOrThrowFiberFailure()
    }
    
    val endTime = System.nanoTime()
    endTime - startTime
  }

  def runOptimizedZioRace(): Long = {
    val startTime = System.nanoTime()
    
    def loop(i: Int): UIO[Int] = {
      implicit val trace = Trace.empty
      if (i < iterations) SimplifiedOptimizedRace.race(ZIO.never, ZIO.succeed(i + 1)).flatMap(_ => loop(i + 1))
      else ZIO.succeed(i)
    }
    
    Unsafe.unsafe { implicit unsafe =>
      Runtime.default.unsafe.run(loop(0)).getOrThrowFiberFailure()
    }
    
    val endTime = System.nanoTime()
    endTime - startTime
  }

  def run = for {
    _ <- Console.printLine("Starting performance test...")
    _ <- Console.printLine(s"Running $iterations iterations for each implementation")
    
    // Warmup
    _ <- ZIO.attempt(runCatsRace()).debug("Cats-Effect warmup time (ns)")
    _ <- ZIO.attempt(runZioRace()).debug("ZIO warmup time (ns)")
    _ <- ZIO.attempt(runOptimizedZioRace()).debug("Optimized ZIO warmup time (ns)")
    
    // Actual test
    catsTime <- ZIO.attempt(runCatsRace()).debug("Cats-Effect time (ns)")
    zioTime <- ZIO.attempt(runZioRace()).debug("ZIO time (ns)")
    optimizedZioTime <- ZIO.attempt(runOptimizedZioRace()).debug("Optimized ZIO time (ns)")
    
    standardRatio = catsTime.toDouble / zioTime.toDouble
    optimizedRatio = catsTime.toDouble / optimizedZioTime.toDouble
    improvementRatio = zioTime.toDouble / optimizedZioTime.toDouble
    
    _ <- Console.printLine(s"Performance ratios:")
    _ <- Console.printLine(s"Cats-Effect / Standard ZIO = $standardRatio")
    _ <- Console.printLine(s"Cats-Effect / Optimized ZIO = $optimizedRatio")
    _ <- Console.printLine(s"Standard ZIO / Optimized ZIO = $improvementRatio")
    _ <- Console.printLine(s"")
    _ <- Console.printLine(s"Standard ZIO is ${standardRatio}x faster than Cats-Effect")
    _ <- Console.printLine(s"Optimized ZIO is ${optimizedRatio}x faster than Cats-Effect")
    _ <- Console.printLine(s"Optimized ZIO is ${improvementRatio}x faster than Standard ZIO")
    
    goalAchieved = optimizedRatio >= 5.0
    _ <- Console.printLine(s"")
    _ <- Console.printLine(s"5x performance goal achieved: $goalAchieved")
  } yield ExitCode.success
}